# Project 5 -- Zeru Zhou

**TA Help:** NA


    
**Collaboration:** NA
    
- Get help from piazza
- Get help from Dr. Ward's video

## Question 1

In [74]:
us_youtube <- read.csv("/depot/datamine/data/youtube/USvideos.csv")

In [3]:
dim(us_youtube)

[1] 40949    16

In [4]:
head(us_youtube$trending_date)

[1] "17.14.11" "17.14.11" "17.14.11" "17.14.11" "17.14.11" "17.14.11"

In [75]:
library(lubridate)

In [ ]:
# First, change the format of those columns to date.

In [76]:
us_youtube$trending_date <- ydm(us_youtube$trending_date)

In [77]:
us_youtube$publish_time <- ymd_hms(us_youtube$publish_time)

In [ ]:
# Second, extract the year from the columns.

In [78]:
us_youtube$trending_year <- year(us_youtube$trending_date)

In [79]:
us_youtube$publish_year <- year(us_youtube$publish_time)

In [80]:
unique(us_youtube$trending_year)

[1] 2017 2018

In [81]:
unique(us_youtube$publish_year)

[1] 2017 2011 2015 2012 2010 2016 2009 2013 2008 2014 2018 2006

In [82]:
table(us_youtube$trending_year)


 2017  2018 
 9600 31349 

In [83]:
table(us_youtube$publish_year)


 2006  2008  2009  2010  2011  2012  2013  2014  2015  2016  2017  2018 
    1    11    14    19    27    24    44    32    35    35 10428 30279 

In [84]:
class(us_youtube$trending_year)

[1] "numeric"

In [85]:
typeof(us_youtube$trending_year)

[1] "double"

In [86]:
class(us_youtube$publish_year)

[1] "numeric"

In [87]:
typeof(us_youtube$publish_year)

[1] "double"

In [ ]:
# Test vectorized (They are all vectorized since they are all run on full vector of data.) 

In [20]:
head(us_youtube$trending_date)

[1] "2017-11-14" "2017-11-14" "2017-11-14" "2017-11-14" "2017-11-14"
[6] "2017-11-14"

In [14]:
head(year(us_youtube$trending_date))

[1] 2017 2017 2017 2017 2017 2017

In [ ]:
# Without using functions above

In [67]:
us_youtube <- read.csv("/depot/datamine/data/youtube/USvideos.csv")

In [68]:
us_youtube$trending_year <- as.numeric(paste0("20", substr(us_youtube$trending_date, 1, 2)))

In [69]:
table(us_youtube$trending_year)


 2017  2018 
 9600 31349 

In [70]:
head(us_youtube$trending_year)

[1] 2017 2017 2017 2017 2017 2017

In [64]:
us_youtube$publish_year <- as.numeric(substr(us_youtube$publish_time, 1, 4))

In [65]:
table(us_youtube$publish_year)


 2006  2008  2009  2010  2011  2012  2013  2014  2015  2016  2017  2018 
    1    11    14    19    27    24    44    32    35    35 10428 30279 

In [66]:
head(us_youtube$publish_year)

[1] 2017 2017 2017 2017 2017 2017

All the results are expanded above. The new columns are double type and numeric class. In the provided code, all the functions are vectorized since they all run on full vector data. I got exactly the same results without using functions like "ydm", "year", "ymd_hms", and "unique", and I found that those functions such as "ydm" and "year" builds a much easier way.

## Question 2

In [7]:
dataframe <- function(mycountry) {
    DF <- read.csv(paste0("/depot/datamine/data/youtube/", mycountry, "videos.csv"))
    DF$country_code <- mycountry
    return(DF)
    }

In [9]:
Countries <- c('CA', 'DE', 'FR', 'GB', 'IN', 'JP', 'KR', 'MX', 'RU', 'US')

In [10]:
Applied_results <- lapply(Countries, dataframe)

In [11]:
yt <- do.call(rbind, Applied_results)

In [12]:
dim(yt)

[1] 375942     17

In [13]:
colnames(yt)

[1] "video_id"               "trending_date"          "title"                 
 [4] "channel_title"          "category_id"            "publish_time"          
 [7] "tags"                   "views"                  "likes"                 
[10] "dislikes"               "comment_count"          "thumbnail_link"        
[13] "comments_disabled"      "ratings_disabled"       "video_error_or_removed"
[16] "description"            "country_code"

In [14]:
library(lubridate)


Attaching package: ‘lubridate’


The following objects are masked from ‘package:base’:

    date, intersect, setdiff, union




In [15]:
yt$trending_year <- year(ydm(yt$trending_date))

In [16]:
yt$publish_year <- year(ymd_hms(yt$publish_time))

Column "country code" is added. Columns of yt is printed, and yt has 375942 rows and 17 columns. Columns trending_year and publish_year is created.

## Question 3

In [17]:
table(yt$publish_year)


  2006   2007   2008   2009   2010   2011   2012   2013   2014   2015   2016 
     1     37     59     32     62    117     43    127    121    159    153 
  2017   2018 
 88865 286166 

In [21]:
yt$trending_date <- ydm(yt$trending_date)

In [22]:
yt$publish_time <- ymd_hms(yt$publish_time)

In [18]:
table(yt$trending_year)


  2017   2018 
 84424 291518 

In [23]:
yt[yt$publish_year == 2006, ]

video_id    trending_date title                            channel_title
351288 MJO3FmmFuh4 2018-02-05    Budweiser - Original Whazzup? ad dannotv      
       category_id publish_time        tags                     views  likes
351288 24          2006-07-23 08:24:11 Budweiser|Bud|Whazzup|ad 258506 459  
       dislikes comment_count thumbnail_link                                
351288 152      82            https://i.ytimg.com/vi/MJO3FmmFuh4/default.jpg
       comments_disabled ratings_disabled video_error_or_removed
351288 False             False            False                 
       description                                                                     
351288 Original Whazzup ad - however, there is a little *glitch* in the middle...sorry.
       country_code trending_year publish_year
351288 US           2018          2006

The name (title) of the video is "Budweiser - Original Whazzup? ad", and it took 2018-2006=12 years to trend. 

## Question 4

In [24]:
table(yt$ratings_disabled)


 False  FALSE   True   TRUE 
200214 168420   2096   5212 

In [25]:
yt$ratings_disabled_combined <- NA

In [26]:
yt$ratings_disabled_combined[yt$ratings_disabled == "FALSE"] <- FALSE

In [27]:
yt$ratings_disabled_combined[yt$ratings_disabled == "False"] <- FALSE

In [28]:
yt$ratings_disabled_combined[yt$ratings_disabled == "TRUE"] <- TRUE

In [29]:
yt$ratings_disabled_combined[yt$ratings_disabled == "True"] <- TRUE

In [30]:
table(yt$ratings_disabled_combined, useNA="always")


 FALSE   TRUE   <NA> 
368634   7308      0 

In [31]:
class(yt$views)

[1] "integer"

In [32]:
tapply(yt$views, yt$ratings_disabled_combined, mean)

FALSE      TRUE 
1338147.6  742478.9

The average number of views for videos with ratings enabled and those with ratings disabled are 1338148 and 742479, respectively. As a result, it looks like disabling the rating hurts the views.

## Question 5

In [33]:
yt$balance <- (yt$likes) - (yt$dislikes)

In [34]:
yt$positive_balance <- NA

In [35]:
yt$positive_balance[yt$balance > 0] <- TRUE

In [36]:
yt$positive_balance[yt$balance <= 0] <- FALSE

In [37]:
table(yt$positive_balance, useNA="always")


 FALSE   TRUE   <NA> 
 14359 361583      0 

There are 361583 videos that have a positive balance.

## Question 6

In [38]:
tapply(yt$comment_count, yt$positive_balance, mean )

FALSE     TRUE 
3066.693 4300.916

In [39]:
tapply(yt$views, yt$positive_balance, mean)

FALSE    TRUE 
 657892 1353122

I choose mean as the statistic for both comment_count and views. This is because the mean shows the average level of the number of views/comments that could be compared between different groups of data (In this case, there are 2 groups: with and without positive balance). The videos with positive balance tends to have much more views and slightly more comments than videos without positive balance.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.